ADX-381 Fix accepted member requests 404-ing

Manoj-nathwani opened the pull request:
# Problem
- When a membership request has already been actioned and the link is revisited again the page returns a 404
- It gives the impression as if something has gone wrong with the request
- It's also common for other admins within the same org to get an email to the link too, so this is a very common issue

![image](https://user-images.githubusercontent.com/2634482/95069789-17c2c680-06ff-11eb-83de-6b27acc1cfe2.png)


# Solution
- Page no longer returns a 404
- We are clearly explaining that no further action is required
- The status of the membership is also displayed so it's clear if it was `accepted/rejected`

![image](https://user-images.githubusercontent.com/2634482/95070026-6d976e80-06ff-11eb-9d09-c7135da6d868.png)

Commit:
ADX-381 Fix accepted member requests 404-ing

jonathansberry commented on ckanext/ytp/request/templates/request/show.html:
```diff
@@ -42,31 +42,38 @@ <h1>{% trans %}Organization membership request{% endtrans %}</h1>
+          </div>
+        {% else %}
+          <div>
+            <span>This request has already been actioned</span>
```
We need to make sure these strings are picked up by our translation system. 

To do this wrap them in {{_('')}}.  The underscore function is shorthand for our internationalisation function. e.g...

```<span>{{_('This request has already been actioned')}}</span>```

jonathansberry replied:
The other way of doing it is: ```{% trans %}This request has already been actioned{% endtrans %}```

jonathansberry commented on ckanext/ytp/request/templates/request/show.html:
```diff
@@ -42,31 +42,38 @@ <h1>{% trans %}Organization membership request{% endtrans %}</h1>
               <tr>
                   <th>{% trans %}State{% endtrans %}</th>
-                  <td>{{ membership.state }}</td>
+                  <td>{{ membership.state|capitalize }}</td>
```
:ok_hand:

jonathansberry commented on ckanext/ytp/request/logic/action/get.py:
```diff
@@ -30,7 +30,7 @@ def member_request(context, data_dict):
     member_dict['group_id'] = membership.group_id
     member_dict['role'] = member_request.role
-    member_dict['state'] = 'pending'
+    member_dict['state'] = member_request.status
```
This seems really wierd to me.  It seems like such an obvious fault on the part of the original coders that I almost wonder if we are missing something. Anyhow time will tell...

tomeksabala replied:
It the old logic this was the only state the member_request would be in. Otherwise the code was raising a logic exception NotFound.
The other place the issue could be fixed could be:
`ckanext-ytp-request/ckanext/ytp/request/controller.py:72`
```
    def show(self, mrequest_id):
        """"
        Shows a single member request. To be used by admins in case they
        want to modify granted role or accept via e-mail.
        """
        context = {'user': c.user or c.author}
        try:
            membershipdto = toolkit.get_action('member_request_show')(
                context, {'mrequest_id': mrequest_id})
            member_user = model.Session.query(
                model.User).get(membershipdto['user_id'])
            context = {'user': member_user.name}
            roles = self._get_available_roles(
                context, membershipdto['organization_name'])
            extra_vars = {"membership": membershipdto,
                          "member_user": member_user, "roles": roles}
            return render('request/show.html', extra_vars=extra_vars)
        except logic.NotFound:
            abort(404, self.request_not_found_message)
        except logic.NotAuthorized:
            abort(401, self.not_auth_message)
````
Where we could just render a different template to display message "Member request already accepted or not found".

jonathansberry reviewed (changes requested):
Yeah looks like a much more simple fix than I was expecting.  Great job fixing it.  We just need to translate the strings and create a regression test that will: 
- Create an actioned membership request.
- Load the membership page.
- Check that no http error is returned and maybe that "This request has already been actioned" is in the response contents.

tomeksabala reviewed (approved):
This looks good. I was also thinking that it could be fixed by changing the behaviour of
`ckanext-ytp-request/ckanext/ytp/request/controller.py:72` `show` method rendering a new template "Member request already accepted or not found".
That would preserve behaviour of function `member_request` raising not found error if no active member request found.

But after reviewing your updated `show.html` template I think that your decision is the way to go.
Resolve Jonathan suggestions about translations and we're good to go :100:

Manoj-nathwani committed:
added translations

Manoj-nathwani committed:
fix random error in existing plugin

Manoj-nathwani committed:
added tests

Manoj-nathwani commented on ckanext/ytp/request/tests/test_plugin.py:
```diff
@@ -1,44 +1,65 @@
+    _load_plugins = (u'ytp_request', )
+
+    @mock.patch('ckanext.ytp.request.logic.action.create.flash_success')
+    @mock.patch('ckanext.ytp.request.logic.action.update.flash_success')
```
@tomeksabala can you confirm this is a good way to mock?

tomeksabala replied:
I think you could get rid of the mocks if you'd use the test app to call the actions. Example:
```
        app = helpers._get_test_app()

        with app.flask_app.test_request_context():
            app.get(
                url_for(
                    controller='group',
                    action='member_new',
                    id=group['id'],
                ),
                status=403,
            
```
But we're dealing with pylons here so I'm not sure how exactly to do that. Let's leave it right now.

Manoj-nathwani commented on ckanext/ytp/request/tests/test_plugin.py:
```diff
@@ -1,44 +1,65 @@
+            status=200
+        )
+
+        # app.get() will raise a "Bad response" error if the page 404s
```
@tomeksabala can you also confirm this is ok. I don't need a try/catch do I?

Manoj-nathwani replied:
Done ✅

Manoj-nathwani commented:
# Note
You'll need the latest version of adx_develop as i've made the following update: https://github.com/fjelltopp/adx_develop/commit/a4e08ee34fe418daa42e82f4a7f2106e6d0860fa

Manoj-nathwani committed:
updated function name

tomeksabala commented on ckanext/ytp/request/tests/test_plugin.py:
```diff
@@ -1,44 +1,65 @@
+            status=200
+        )
+
+        # app.get() will raise a "Bad response" error if the page 404s
```
Yes, I think you need a proper "assert" section here. It quite unique scenario but the following should do the job:
```
    try:
        app.get(
            url,
            extra_environ={'REMOTE_USER': sysadmin['name'].encode('ascii')},
            status=200
        )
    except BadResponseError:
        self.fail("meaningful test fail message which'll help find out what's wrong in the future")
```
more info (take a look at the most voted comment to this answer): https://stackoverflow.com/a/4319870/3432120

tomeksabala reviewed (changes requested):
This looks good. I'd however update the "assert" part of the test so it fails (instead of errors) if the test doesn't pass.
Please see my inline comment from the review.

Manoj-nathwani committed:
better error handling

tomeksabala commented on ckanext/ytp/request/tests/test_plugin.py:
```diff
@@ -56,10 +56,11 @@ def test_viewing_actioned_referral(self, mocked_create_flash, mocked_update_flas
+                extra_environ={'REMOTE_USER': sysadmin['name'].encode('ascii')},
+                status=200
+            )
+        except BadResponseError:
```
That's not going to work. `BadResponseError` is undefined in your code. Travis has caught that as well:
https://travis-ci.org/github/fjelltopp/ckanext-ytp-request/jobs/736091133

tomeksabala commented on ckanext/ytp/request/tests/test_plugin.py:
```diff
@@ -56,10 +56,11 @@ def test_viewing_actioned_referral(self, mocked_create_flash, mocked_update_flas
+                status=200
+            )
+        except BadResponseError:
+            self.fail('Membership request failed to display after being actioned')
```
Could you add the new line at the end of the file too?

Manoj-nathwani committed:
Fix error handling of test

TEMP - break travis to confirm error works

TEMP - revert last commit

TEMP - break travis to confirm error works

TEMP - revert last commit

added initdb to travis config

fixed travis config

added email mocking

fixed pep8 errors

Manoj-nathwani commented:
@tomeksabala - I've done a git squash & changed the author to myself. So this is 100% done and ready to be merged 🚀
